### PR TITLE
feat: enable multiple signature schemas

### LINKED ISSUE #22033: Enable mutliple signature schemes for event signing
Depends on #22010 

With the above PR merged, event creator can use different signing algorithms. This needs to be implemented for event verification as well. Apart from that, we also need method of generating certificates with different key types. Once that is done, the whole system should be able to run with a different signing schema.

### PR DESCRIPTION
**Pull request by lpetrovic05**
closes #22033

### CODE REVIEW

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java [hunk 1]
```diff
@@ -27,10 +27,8 @@ private SigningFactory() {}
     /**
      * The default implementations to use for each schema.
      */
-    private static final Map<String, SigningImplementation> defaultImplementations = Map.of(
-            SigningSchema.RSA.getKeyType(), SigningImplementation.RSA_BC,
-            SigningSchema.EC.getKeyType(), SigningImplementation.EC_JDK,
-            SigningSchema.ED25519.getKeyType(), ED25519_SODIUM);
+    private static final Map<SigningSchema, SigningImplementation> defaultImplementations =
+            Map.of(SigningSchema.RSA, SigningImplementation.RSA_BC, SigningSchema.ED25519, ED25519_SODIUM);
```
Comment (poulok) [inline platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java]: Why is EC removed?

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/internal/SodiumSigner.java [hunk 2]
```diff
@@ -19,11 +19,6 @@ public class SodiumSigner implements BytesSigner {
      * @param keyPair the Ed25519 KeyPair to use for signing
      */
     public SodiumSigner(@NonNull final KeyPair keyPair) {
-        if (!keyPair.getPrivate().getAlgorithm().equals("EdDSA")) {
-            throw new IllegalArgumentException("SodiumSigner only supports EdDSA keys, the supplied key is of type: "
```
Comment (poulok) [inline platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/internal/SodiumSigner.java:23]: Is this no longer true?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java [hunk 3]
```diff
@@ -160,11 +159,6 @@ private static boolean isTransactionNull(@Nullable final Bytes transaction) {
      */
     private boolean areByteFieldsCorrectLength(@NonNull final PlatformEvent event) {
         final GossipEvent gossipEvent = event.getGossipEvent();
-        if (gossipEvent.signature().length() != SignatureType.RSA.signatureLength()) {
```
Comment (poulok) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java:163]: Instead of removing this check, we could check if the signature length is greater than the largest possible signature length given all the types. Do you think that would be useful?

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java (hunk 1 above)
Comment (lpetrovic05) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java]: For the sake of simplicity, we don't plan on using it, we can add it if we need it.

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/internal/SodiumSigner.java (hunk 2 above)
Comment (lpetrovic05) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/internal/SodiumSigner.java:23]: it is, there is just a bit of complication. I moved this check to a static method in SigningSchema, and removed it from here to avoid checking it twice. we could check it multiple times, but some of the current uses of this create a new instance every time, so I wanted to avoid it. do you think i should put it back in?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java (hunk 3 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java:163]: I thought about that option, and its definitely a possibility. But the more I thought about it, the more I don't see what this check actually gives us. If a node sends a big signature, the signature check will fail, then we will punish this node for sending us bad events. I don't see any value in having this check.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/standalone/TransactionExecutorsTest.java [hunk 4]
```diff
@@ -605,7 +605,8 @@ public static X509Certificate randomX509Certificate() {
             final KeyPair rsaKeyPair1 = rsaKeyGen.generateKeyPair();
 
             final String name = "CN=Bob";
-            return CryptoStatic.generateCertificate(name, rsaKeyPair1, name, rsaKeyPair1, secureRandom);
+            return CryptoStatic.generateCertificate(
+                    name, rsaKeyPair1, name, rsaKeyPair1, secureRandom, "SHA384withRSA");
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/standalone/TransactionExecutorsTest.java]: Nit: you could use of the constants here

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java [hunk 5]
```diff
@@ -2,97 +2,39 @@
...
-            Objects.requireNonNull(keysAndCerts, "keysAndCerts must not be null");
-            final Signature s = Signature.getInstance(CryptoConstants.SIG_TYPE2, CryptoConstants.SIG_PROVIDER);
-            s.initSign(keysAndCerts.sigKeyPair().getPrivate());
-            this.signature = s;
-        } catch (final NoSuchAlgorithmException | NoSuchProviderException | InvalidKeyException e) {
-            throw new PlatformConstructionException(e);
-        }
+        this.signer = SigningFactory.createSigner(keysAndCerts.sigKeyPair());
+        this.signatureType = switch (SigningSchema.fromKeyType(
+                keysAndCerts.sigKeyPair().getPrivate())) {
```
Comment (netopyr) [inline platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java]: Nit: I'd store `SigningSchema.fromKeyType(keysAndCerts.sigKeyPair().getPrivate())` in a local variable

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java [hunk 6]
```diff
@@ -2,97 +2,39 @@
...
+            case RSA -> SignatureType.RSA;
+            case ED25519 -> SignatureType.ED25519;
+        };
```
Comment (netopyr) [inline platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java:29]: Looking at this mapping, I wonder if it would make sense to merge `SignatureType` and `SigningSchema` or make one part of the other.

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java (hunk 1 above)
Comment (netopyr) [inline platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java]: `EnumMap` is slightly more efficient.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java (hunk 3 above)
Comment (netopyr) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java:163]: The value would be that we could skip a few steps if we fail here.

Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java:163]: It would be adding a step to skip a step for one event in a million. And we would only be able to catch if the signature is bigger than an RSA signature. I feel like less is more, and adding a step for an edge case is a bigger cost then it is a benefit.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/standalone/TransactionExecutorsTest.java (hunk 4 above)
Comment (lpetrovic05) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/standalone/TransactionExecutorsTest.java]: done

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java (hunk 5 above)
Comment (lpetrovic05) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java]: yeah, its a lot more readable that way

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java (hunk 6 above)
Comment (lpetrovic05) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java:29]: I thought about this, and decided against it for the following reasons:
- SignatureType is used to be serialized as part of a signature. This uses the old serialization mechanism that we want to get rid of. In protobuf, we don't serialize the sig type, just the bytes.
- It has a field (provider) that specifies the implementation, which I think should be a separate concept.
- It seems to be used mostly by things that we are hopefully about to delete:
  - PTT (should be deleted once JRS is gone)
  - CES (should be deleted once we block streams go live)
  - Record streams (again, not needed once block streams go live)
  
I'm can reconsider if you think we should combine the two.

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java (hunk 1 above)
Comment (lpetrovic05) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/SigningFactory.java]: did not know this exists, updated

**Review by poulok — APPROVED** [review]

File: platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java (hunk 6 above)
Comment (netopyr) [reply platform-sdk/consensus-utility/src/main/java/org/hiero/consensus/crypto/PlatformSigner.java:29]: No, I agree with you. If it's planned to phase out this enum and this is the only place where we do this mapping, it's probably not worth it.

**Review by netopyr — APPROVED** [review]
LGTM - thanks @lpetrovic05

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java (hunk 3 above)
Comment (netopyr) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/validation/DefaultInternalEventValidator.java:163]: The question to ask is whether this could be used in an attack, in which case it would not be a one-in-a-million occurrence. Anyways, we would be able to skip deduplication only where the size of the signature does not matter. I agree. It probably isn't worth it.
